extension/observer/cfgardenobserver/extension.go

```diff
@@ -4,33 +4,337 @@
+		g.apps[appID] = app
+	}
+
+	return nil
+}
+
+func (g *cfGardenObserver) App(info garden.ContainerInfo) (*resource.App, error) {
+	appID, err := infoToAppID(info)
+	if err != nil {
+		return nil, err
+	}
+
+	g.appMu.Lock()
+	defer g.appMu.Unlock()
+	app, ok := g.apps[appID]
+	if ok {
+		return app, nil
+	}
+
+	app, err = g.cf.Applications.Get(g.ctx, appID)
+	if err != nil {
+		return nil, err
+	}
+	g.apps[appID] = app
+
+	return app, nil
+}
+
+func (g *cfGardenObserver) Start(_ context.Context, _ component.Host) error {
+	gCtx, cancel := context.WithCancel(context.Background())
+	g.cancel = cancel
+	g.ctx = gCtx
+
+	g.garden = gardenClient.New(gardenConnection.New("unix", g.config.Garden.Endpoint))
+
+	var err error
+	g.cf, err = newCfClient(g.config.CloudFoundry)
+	if err != nil {
+		return err
+	}
+
+	if g.config.IncludeAppLabels {
+		g.once.Do(
+			func() {
+				go func() {
+					cacheRefreshTicker := time.NewTicker(g.config.CacheSyncInterval)
+					defer cacheRefreshTicker.Stop()
+
+					for {
+						select {
+						case <-g.ctx.Done():
+							return
+						case <-cacheRefreshTicker.C:
+							err = g.SyncApps()
+							if err != nil {
+								g.logger.Error("could not sync app cache", zap.Error(err))
+							}
+						}
+					}
+				}()
+			},
+		)
+	}
+
+	return nil
+}
+
+func (g *cfGardenObserver) Shutdown(_ context.Context) error {
+	g.cancel()
+	return nil
+}
+
 func (g *cfGardenObserver) ListEndpoints() []observer.Endpoint {
-	// TODO: Implement the logic to list the endpoints.
-	endpoints := make([]observer.Endpoint, 0)
+	var endpoints []observer.Endpoint
+
+	containers, err := g.garden.Containers(garden.Properties{})
+	if err != nil {
+		g.logger.Error("could not list containers", zap.Error(err))
+		return endpoints
+	}
+
+	infos := make(map[string]garden.ContainerInfo)
+	for _, c := range containers {
+		info, err := c.Info()
+		if err != nil {
+			g.logger.Error("error getting container info", zap.String("handle", c.Handle()), zap.Error(err))
+			continue
+		}
+
+		if info.State != "active" {
+			continue
+		}
+
+		endpoints = append(endpoints, g.containerEndpoints(c.Handle(), info)...)
+		infos[c.Handle()] = info
+	}
 
+	go g.updateContainerCache(infos)
 	return endpoints
 }
+
+// containerEndpoints generates a list of observer.Endpoint for a container,
+// this is because a container might have more than one exposed ports
+func (g *cfGardenObserver) containerEndpoints(handle string, info garden.ContainerInfo) []observer.Endpoint {
+	portsProp, ok := info.Properties["network.ports"]
+	if !ok {
+		g.logger.Error("could not discover container ports")
+		return nil
+	}
+	ports := strings.Split(portsProp, ",")
+
+	var app *resource.App
+	var err error
+	if g.config.IncludeAppLabels {
+		app, err = g.App(info)
+		if err != nil {
+			g.logger.Error("error fetching Application", zap.Error(err))
+			return nil
+		}
+	}
+
+	endpoints := []observer.Endpoint{}
+	for _, port := range ports {
+		port, err := strconv.ParseUint(port, 10, 16)
```

Comment:
`err` is already declared, so we don't need the `:=`, maybe we declare `port` above and convert this to a plain `=`